src/baldr/graphreader.cc

```diff
@@ -37,31 +37,61 @@ tile_gone_error_t::tile_gone_error_t(std::string prefix, baldr::GraphId edgeid)
 
 GraphReader::tile_extract_t::tile_extract_t(const boost::property_tree::ptree& pt) {
   // if you really meant to load it
-  if (pt.get_optional<std::string>("tile_extract")) {
+  auto extract_fp = pt.get_optional<std::string>("tile_extract");
+  if (extract_fp) {
     try {
-      // load the tar
-      archive.reset(new midgard::tar(pt.get<std::string>("tile_extract")));
-      // map files to graph ids
-      for (auto& c : archive->contents) {
-        try {
-          auto id = GraphTile::GetTileId(c.first);
-          tiles[id] = std::make_pair(const_cast<char*>(c.second.first), c.second.second);
-        } catch (...) {
-          // It's possible to put non-tile files inside the tarfile.  As we're only
-          // parsing the file *name* as a GraphId here, we will just silently skip
-          // any file paths that can't be parsed by GraphId::GetTileId()
-          // If we end up with *no* recognizable tile files in the tarball at all,
-          // checks lower down will warn on that.
+      // map files to graph ids: use precomputed index.bin if available, else scan the tar
+      std::ifstream fin(*extract_fp, std::ios::binary);
+
+      tar::header_t first_header;
+      fin.read((char*)&first_header, sizeof(tar::header_t));
+      if (first_header.verify() &&
+          std::strstr(const_cast<char*>(first_header.name), "valhalla_tiles/index.bin")) {
+        fin.seekg(sizeof(tar::header_t));
+        auto index_size = first_header.get_file_size();
+        auto block_size = sizeof(tar::header_t);
+        auto remainder = index_size % block_size;
+        // the initial data position is header + index_size + padding
+        auto tile_pos =
+            block_size + index_size + block_size - (remainder == 0 ? block_size : remainder);
+        while (fin.tellg() != first_header.get_file_size() + block_size) {
+          uint32_t tile_id = 0;
+          uint64_t tile_padded_size = 0;
+          fin.read(reinterpret_cast<char*>(&tile_id), sizeof(tile_id));
+          fin.read(reinterpret_cast<char*>(&tile_padded_size), sizeof(tile_padded_size));
+
+          // add header size to position, subtract header size from file size
+          tiles[tile_id] = std::make_pair(reinterpret_cast<char*>(&tile_pos + block_size),
```

Comment:
tile_id & tile_size are correctly set, `tile_pos` is not. missing the right type casting..

also, same problem with endianess etc as in the python equivalent: https://github.com/valhalla/valhalla/pull/3281/files#r694160065.

any suggestions here would be welcome!